llvm-external-projects/iree-dialects/lib/Transforms/TransformMatchers.cpp

```diff
@@ -952,26 +972,40 @@ void transform_ext::makeSoftmaxMatcher(
             .output(AllOperands(), IsProjected(-1))
             .output(NumEqualsTo(1));
   sum = sum.input(0, expOperand);
-  sum = sum.output(0, fillzero);
+  sum = sum.output(0, fillZero);
 
-  divOperand = m_StructuredOp<linalg::GenericOp>()
+  rcpOperand = m_StructuredOp<linalg::GenericOp>()
                    .isFloatReciprocal()
                    .dim(AllDims(), utils::IteratorType::parallel)
                    .input(NumEqualsTo(1))
                    .input(AllOperands(), IsIdentity())
                    .output(AllOperands(), IsIdentity())
                    .output(NumEqualsTo(1));
-  divOperand = divOperand.input(0, sum);
-
-  softmaxroot = transform_ext::m_StructuredOp<linalg::GenericOp>()
-                    .singleOpWithCanonicaleArgs<arith::MulFOp>()
-                    .dim(AllDims(), utils::IteratorType::parallel)
-                    .input(NumEqualsTo(2))
-                    .input(0, IsIdentity())
-                    .input(1, IsProjected(-1))
-                    .output(NumEqualsTo(1))
-                    .output(AllOperands(), IsIdentity());
-
-  softmaxroot = softmaxroot.input(0, expOperand);
-  softmaxroot = softmaxroot.input(1, divOperand);
+  rcpOperand = rcpOperand.input(0, sum);
+
+  mulOperand = transform_ext::m_StructuredOp<linalg::GenericOp>()
+                   .singleOpWithCanonicaleArgs<arith::MulFOp>()
+                   .dim(AllDims(), utils::IteratorType::parallel)
+                   .input(NumEqualsTo(2))
+                   .input(0, IsIdentity())
+                   .input(1, IsProjected(-1))
+                   .output(NumEqualsTo(1))
+                   .output(AllOperands(), IsIdentity());
+
+  mulOperand = mulOperand.input(0, expOperand);
+  mulOperand = mulOperand.input(1, rcpOperand);
+
+  divOperand = transform_ext::m_StructuredOp<linalg::GenericOp>()
+                   .singleOpWithCanonicaleArgs<arith::DivFOp>()
+                   .dim(AllDims(), utils::IteratorType::parallel)
+                   .input(NumEqualsTo(2))
+                   .input(0, IsIdentity())
+                   .input(1, IsProjected(-1))
+                   .output(NumEqualsTo(1))
+                   .output(AllOperands(), IsIdentity());
+
+  divOperand = divOperand.input(0, expOperand);
+  divOperand = divOperand.input(1, sum);
+
+  softmaxroot = transform_ext::m_StructuredOp_Or(mulOperand, divOperand);
```

Comment:
Nice! We can also copy the matchers, so something like

```
auto softmaxRootBase = transform_ext::m_StructuredOp<linalg::GenericOp>()
                   .dim(AllDims(), utils::IteratorType::parallel)
                   .input(NumEqualsTo(2))
                   .input(0, IsIdentity())
                   .input(1, IsProjected(-1))
                   .output(NumEqualsTo(1))
                   .output(AllOperands(), IsIdentity())
                   .input(0, expOperand);

auto mulOperandCopy = softmaxRootBase;
auto divOperandCopy = softmaxRootBase;
mulOperand = mulOperandCopy.input(1, rcpOperand);
divOperand = divOperandCopy.input(1, sum);
```

will remove the repetition. It's annoying because of the current ownership scheme, I'll look into that.